pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -447,17 +510,59 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListenerForPortOrUDS(no
 		// add one empty entry to the list so we generate a default listener below
 		allChains = []plugin.FilterChain{{}}
 	}
-	for _, chain := range allChains {
+
+	// Detect protocol by sniffing and double the filter chain
+	if pluginParams.ListenerProtocol == plugin.ListenerProtocolAuto {
+		allChains = append(allChains, allChains...)
+		listenerOpts.needHTTPInspector = true
+	}
+
+	for id, chain := range allChains {
 		var httpOpts *httpListenerOpts
 		var tcpNetworkFilters []*listener.Filter
+		var filterChainMatch *listener.FilterChainMatch
 
 		switch pluginParams.ListenerProtocol {
 		case plugin.ListenerProtocolHTTP:
+			filterChainMatch = chain.FilterChainMatch
 			httpOpts = configgen.buildSidecarInboundHTTPListenerOptsForPortOrUDS(node, pluginParams)
 
 		case plugin.ListenerProtocolTCP:
+			filterChainMatch = chain.FilterChainMatch
 			tcpNetworkFilters = buildInboundNetworkFilters(pluginParams.Env, pluginParams.Node, pluginParams.ServiceInstance)
 
+		case plugin.ListenerProtocolAuto:
+			fcm := listener.FilterChainMatch{}
+			// Build filter chain options for listener configured with protocol sniffing
+			// Double the number of filter chains. Half of filter chains are used as http filter chain and half of them are used as tcp proxy
+			// id in [0, len(allChains)/2) are configured as http filter chain, [(len(allChains)/2, len(allChains)) are configured as tcp proxy
+			// If mTLS is enabled, there are four filter chains. The filter chain match should be
+			//  FCM 1: ALPN [istio, http/1.0, http/1.1, h2] Transport protocol: tls
+			//  FCM 2: ALPN [http/1.0, http/1.1, h2] Transport protocol: N/A
+			//  FCM 3: ALPN [istio] Transport protocol: N/A
```

Comment:
this should be `FCM3: ALPN [istio], Transport protocol: tls` right?

Reply:
No. FCM1 will be a superset of FCM3 if setting transport protocol to `tls` and Envoy will reject the configuration.